coco_keyprovider/README.md

```diff
@@ -1,146 +1,197 @@
 
-The `<parameters>` in skopeo command is a key-value list separated by double colons. Here are the defined keys:
+### The encryption protocol explained
+
+As shown on the previous section, the encryption protocol (`--encryption-key`) passed to skopeo has the `provider:attestation-agent:<parameters>` format.
+
+The `<parameters>` is a key-value list separated by double colons (e.g. key1=value1::key2=value2). Here are the defined keys:
 - `sample`: Not required. Either `true` or `false`. If not set, use `false`. This value indicates whether the hardcoded encryption key is used. This works the same way as `sample keyprovider`.
-- `keyid`: Required if `sample` is not enabled. It is a KBS Resource URI, s.t. `kbs://<kbs-addr>/<repo>/<type>/<tag>`. When decryption occurs, the `keyid` value is used to index the KEK.
-- `keypath`: Required if `sample` is not enabled. A local filesystem path, absolute path recommended. Specify the KEK to encrypted the image in local filesystem. KEK will be read from fs and then used to encrypt the image. This key's length must be 32 bytes.
+- `keyid`: Required if `sample` is not enabled. It is a Key Broker Service (KBS) Resource URI, s.t. `kbs://<kbs-addr>/<repo>/<type>/<tag>`. When decryption occurs, the `keyid` value is used to index the Key Encryption Key (KEK).
+- `keypath`: Required if `sample` is not enabled. A local filesystem path, absolute path recommended. Specify the KEK to encrypt the image in local filesystem. KEK will be read from filesystem and then used to encrypt the image. This key's length must be 32 bytes.
 - `algorithm`: Not required. Indicate the encryption algorithm used. Either `A256GCM` or `A256CTR`. If not provided, use `A256GCM` by default as it is AEAD scheme.
 
 ### Examples
 
-Generate a key to encrypt the image
-```bash
-head -c32 < /dev/random > key1
-```
+This section contain encrypting examples.
 
-- Use key of path `key1`, and keyid `kbs:///default/key/key_id1` to encrypt an image. In this way sample is disabled, and will use A256GCM (AES-256-GCM).
-```
-OCICRYPT_KEYPROVIDER_CONFIG=ocicrypt.conf skopeo copy --insecure-policy --encryption-key provider:attestation-agent:keypath=$(pwd)/key1::keyid=kbs:///default/key/key_id1::algorithm=A256GCM docker://busybox oci:busybox:encrypted
+#### Example 1: encrypting for sample kbc
+
+Let's start with the simplest example possible, which is to encrypt an image using the sample key provider:
+
+```shell
+$ skopeo copy --insecure-policy --encryption-key provider:attestation-agent:sample=true docker://busybox oci:busybox_encrypted:sample
 ```
 
-- Use sample key provider to encrypt an image
+#### Example 2: encrypting for Offline fs, Offline SEV and Online SEV KBC
+
+For `offline-fs-kbc`, `offline-sev-kbc` and `online-sev-kbc` the KBS address shoud be set to "default". In common, they need a key upfront to encrypt the image:
```

Comment:
You might want to add a line saying that this key can then be provisioned to the kbs.

Also, I don't think `default` is required.

Reply:
True, `default` is not required but it should have a `<repository>` as the error message below implies:

```
2023-05-15T23:15:52Z DEBUG coco_keyprovider::enc_mods] normalize key id kbs:///key/key_id1
FATA[0004] copying system image from manifest list: Unable to finalize encryption: error while retrieving keyprovider protocol grpc output: Error from grpc method: rpc error: code = Internal desc = encrypt failed: Resource path kbs:///key/key_id1 must follow one of the following formats:
                'kbs:///<repository>/<type>/<tag>'
                'kbs://<kbs-addr>/<repository>/<type>/<tag>'
                '<kbs-addr>/<repository>/<type>/<tag>'
                '/<repository>/<type>/<tag>'
```

Can <repository> be anything, subject to user's preference? Or it will be used by some component to whatever reason and should have a correct value (e.g. be the registry repo where the image will be pushed to)?

Reply:
Looks good. Thanks.